### PR TITLE
Add concepts reference

### PR DESCRIPTION
**Pull request by zombiezen**
This documents the near-future target for this tool, not the tool as it exists today.

Fixes ch-2696

### CODE REVIEW

**Comment by jamesnaftel** [comment]
I like it.

File: docs/concepts.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+## Target
+
+A _target_ is a buildable thing &mdash; it could be a service, a library,
+a binary, etc. A target has four _phases_ (all optional, but at least one must
+be present):
+
+-  Build
+-  Test
+-  Execute
+-  Release
```
Comment (glacials) [inline docs/concepts.md:20]: I'm super on board with being opinionated about these four phases existing as first-class citizens, and not just shoving them all into the "build" cubby 👍

However with _each target_ having each of these four phases, it's unclear to me how we expect customers to draw the lines between them. For example, take a project which in today's world has these targets:
```
seeddb
test
integrationtest
release
```
Plus this "non-target":
```
exec
```

With each target gaining build/test/execute/release, how do we expect this project to split their targets up? The closest I can see is something like
```
seeddb
  build
api
  build
  test
  exec
  release
integration
  test
```

And this seems like it's actually more shoehorned than the current system, not less.

Should we consider build/test/exec/release to be _categories of targets_, such that typing `yb test` runs all targets categorized as tests, `yb exec` runs all targets (e.g. PostgreSQL, Redis, app, etc.) categorized as `exec`, etc.?

As a meta-comment, it would be a good practice to go through some of our customers' build configs and refactor them to how they would look in this new system to reveal any oddness. Happy to pair on some of this for the benefit of creative juices.

File: docs/concepts.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# yb Concepts
+
+## Package
+
+A _package_ is a collection of related files and targets. A package is a
+directory that contains a .yourbase.yml file directly inside it. Files can only
+belong to a single package. A package may contain subpackages in subdirectories,
+but files within those subdirectories belong to the subpackage, not the parent
+package.
```
[the remark this answers is not in the document]
Comment (glacials) [reply docs/concepts.md:9]: Does a file in a non-package subdirectory belong to the package, or is it packageless? e.g.
```
./app
  .yourbase.yml
  a.txt
  b/
    c.txt
```
Does `c.txt` belong to `app`?

If so, are we okay with that relationship being implicitly destroyed if `b/.yourbase.yml` is created?

File: docs/concepts.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+different host.
+
+Programs in the context are guaranteed to have access to all the files in the
+package and the following environment variables:
+
+-  `HOME`
+-  `LOGNAME`
+-  `PATH`
+-  `TZ=UTC`
+-  `USER`
```
Comment (glacials) [inline docs/concepts.md:46]: Can we get brief descriptions of these?

**Review by glacials — CHANGES_REQUESTED** [review]
👍 Solid first steps

File: docs/concepts.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+## Buildpacks
+
+A _buildpack_ installs development tools into a context. It sets up environment
+variables and files before a phase is executed.
+
+## Resources
+
+A _resource_ is a process that runs concurrently with the build. This may be
```
Comment (amyhua) [inline docs/concepts.md:55]: can you elaborate?

also, i may shy away from broadly named names. Something more specific is generally better IMO, because we can run into name pollution with colloqial use, very quickly.

File: docs/concepts.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+A _package_ is a collection of related files and targets. A package is a
+directory that contains a .yourbase.yml file directly inside it. Files can only
+belong to a single package. A package may contain subpackages in subdirectories,
+but files within those subdirectories belong to the subpackage, not the parent
+package.
+
+## Target
+
+A _target_ is a buildable thing &mdash; it could be a service, a library,
+a binary, etc. A target has four _phases_ (all optional, but at least one must
```
Comment (amyhua) [inline docs/concepts.md:14]: (nitpick) --> "A target has 1 or more of the following phases"

File: docs/concepts.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+## Buildpacks
+
+A _buildpack_ installs development tools into a context. It sets up environment
+variables and files before a phase is executed.
+
+## Resources
+
+A _resource_ is a process that runs concurrently with the build. This may be
+another target's execute phase or it may be a Docker container.
```
Comment (amyhua) [inline docs/concepts.md:56]: now that i've read the whole thing --- i think, for kind of a sanity check on the doc as a whole, specifying a diagram of simply (without concepts) what is all required, generally, in the whole build/deployment process, and then, mapping those needs to these concepts and how these concepts get applied, would be a nice sanity check and help me with establishing a clearer model of how these concepts all fit as a whole, to the general CI/CD experience.

**Review by amyhua — CHANGES_REQUESTED** [review]
This is a nice initial outline!

I left a number of comments, mostly asking for clarity before moving on to give my thoughts on them.

I propose also specifying the phases more, and reduce their number to 3 for easier picking for the customer.

File: docs/concepts.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+## Target
+
+A _target_ is a buildable thing &mdash; it could be a service, a library,
+a binary, etc. A target has four _phases_ (all optional, but at least one must
+be present):
+
+-  Build
+-  Test
+-  Execute
```
[the remark this answers is not in the document]
Comment (amyhua) [reply docs/concepts.md:19]: I think that phase-definitions, to specify side-effect-freeness, is nice.
However, it should be obvious to a user, without thinking too much, to determine which phase corresponds to which commands they'd like to store under a build target.

IMO I think either there should be _many_ phases, capturing all sorts of use cases, making it very clear which commands go into which phases, OR there should be very _few_ phases, making wrong choices harder to do, and bucketing examples amongst phases, very clearly.

[the remark this answers is not in the document]
Comment (amyhua) [reply docs/concepts.md:19]: and to play devil's advocate: are we _sure_ that there are _no_ "build" commands, that would be BOTH fitting for the Build phase, _and_ side effect free?

What if a "build" command, that legitimately belongs in that bucket, does cause a side effect? What should a user do, in that case, for us?

TLDR are we sure these phases are "MECE"? ([MECE principle](https://en.wikipedia.org/wiki/MECE_principle))

File: docs/concepts.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+-  Release
+
+Each phase, if present, has a list of one or more commands to run. The build
+phase can declare output files. Output files will be accessible to other targets
+that directly depend on this target.
```
Comment (glacials) [inline docs/concepts.md:24]: (unrelated to Amy's above comments, just on the same line) `build` is explicitly allowed to output files, but `test` is not. What about code coverage output?

**Comment by zombiezen** [comment]
Abandoning this PR for the time being until there's more internal alignment.
